### ISSUE TITLE
Integrating Readium into Existing iOS Workspace

### ISSUE THREAD

**Issue by jimjeffers** [readium/SDKLauncher-iOS#47]
Hey guys,

I was wondering if there is any documentation on incorporating the readium-sdk into an existing iOS project? I have a project using cocoapods so what I've done is setup a git submodule for the Readium-sdk and added the xcode project to the workspace. I'm attempting to drop the source code for the launcher app into my own application for a proof of concept but I'm getting a lot of compiler errors from the LocStr() function used throughout the application.

Any help is appreciated.

**Comment by rkwright** [comment]
@jimjeffers  I don't know of any documentation on this approach per se.  It has been done, but by third parties who didn't provide us any documentation.  Can you provide more detail on the compiler errors (like the output)?

**Comment by jimjeffers** [comment]
@rkwright If I can get it working I will put some together for you guys. I need to understand the setup and the process though. Here is what I did to try and mimic the launcher:

1. Added the readium-sdk and readium-shared-js repos as submodules.
2. Added the source files from the SDK launcher app as an attempt to build a proof of concept. I just want to see if I can launch the epub viewer from within my app. 
3. Added the RDServices folder from within the git submodule so that the source code there is part of the main project. 
4. Updated my header search paths to reflect those in the SDK Launcher example app.

It's important to note that the SDK builds and I can drop in the framework. But the compiler errors occur after I add RDServices and the SDK Launcher's files. That's where I can't get any further. I start getting compiler errors and the app will no longer build. Is there another setting I need to set in my build settings so that the low level C code can compile properly?

Here's a sample of some of the dump of syntax errors that prevented the compile. I'm sure it would raise more if it hadn't hit the limit already:

https://gist.github.com/jimjeffers/00a06c64d789ad265d00

**Comment by jimjeffers** [comment]
@rkwright Sorry I should mention this is iOS not OSX.

**Comment by rkwright** [comment]
@jimjeffers Ah, I bet it is because you aren't invoking 
SDKLauncher-iOS/readium-sdk/Platform/Apple/MakeHeaders.sh
This copies a slew of header files to where the readium-sdk project expects them (I don't like this step but that's another story see issue #140). Not sure offhand where in the XCode project that script is invoked. Take a look in the readium-sdk project.

take a look at Build Script in Build Phases:
cd ${SRCROOT}/../..
sh MakeHeaders.sh Apple

**Comment by jimjeffers** [comment]
@rkwright Correct I saw those. So my setup looks like this:

* MyReader.xcodeproj: Run Script: `./Resources/epubReadingSystem.sh` (I've copied this directory into the project with the MathJax, epubReadingSystem.sh, and epubReadingSystem.js)
* ePub3.xcodeproj: Run Script: `cd ${SRCROOT}/../..; sh MakeHeaders.sh Apple`

The app compiles just fine at this point. The issues I'm describing come up when I add the RDServices directory to my xcode project. Is there a compiler flag I need to set on those sources? I guess it'd be best if the epub3 framework was able to provide those internally so that you could install the framework cleanly. But I'd just like to get it compiling ;)

**Comment by rkwright** [comment]
from @agilshane 
I don't have any suggestions. Jim mentions the epub3 "framework", which I'm not familiar with. He said:

> It's important to note that the SDK builds and I can drop in the framework.

...as well as:

> I guess it'd be best if the epub3 framework was able to provide those internally so that you could install the framework cleanly

I don't know what that framework is. The launcher pulls in the ePub3 project directly as a dependency, not as a framework. Sounds like he's trying a new approach and it's not working out for him. My suggestion would be to model his Xcode project dependencies after the launcher.

@jimjeffers   Can you provide a little more detail?

**Comment by jimjeffers** [comment]
@rkwright @agilshane I'll try to put together a short 1-2 minute screencast to show the steps I took. That should probably make it immediately apparent if I'm going about this in completely the wrong way!

**Comment by jimjeffers** [comment]
@rkwright @agilshane here's a link to a short screencast showing how I attempted to recreate the setup in my own project:
http://sumocreations.s3.amazonaws.com/ReadiumInstall.mp4

Thanks for being so responsive guys. Definitely appreciated.

**Comment by agilshane** [comment]
It looks like at the tail end of the ReadiumInstall.mp4 video that the first error has to do with "No member named 'move'...". One thing to check would be the C++ compiler/linker settings on your main project. In the iOS launcher, C++ Language Dialect is GNU++ 11, and C++ Standard Library is libc++.

**Comment by jimjeffers** [comment]
@agilshane Good call. I went through all of the settings to and changed the C++ settings. To confirm I have the settings matching the launcher project: 

C Language Dialect -> GNU99.
C++ Language Dialect -> GNU++11
C++ Standard Lib -> libc++

I cleared derived data, cleaned, and recompiled but still get a fail due to too many warnings:

Here's a screenshot of the error summary (a bit easier to read):
http://sumocreations.s3.amazonaws.com/Screen%20Shot%202015-03-02%20at%205.21.49%20PM.png

----------------------------------------------------------------------------

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:297:49: Use of undeclared identifier 'nullptr_t'; did you mean 'nullptr'?
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDSpineItem.mm:31:9: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDSpineItem.mm:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/manifest.h:24:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/manifest.h:24:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:12: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:17: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/utility:157:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:157:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__tuple:16:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:

----------------------------------------------------------------------------

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:297:49: Template argument for template type parameter must be a type
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDSpineItem.mm:31:9: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDSpineItem.mm:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/manifest.h:24:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/manifest.h:24:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:12: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:17: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/utility:157:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:157:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__tuple:16:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:296:17: Template parameter is declared here

----------------------------------------------------------------------------

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:507:52: Use of undeclared identifier 'nullptr_t'; did you mean 'nullptr'?
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDSpineItem.mm:31:9: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDSpineItem.mm:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/manifest.h:24:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/manifest.h:24:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:12: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:17: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/utility:157:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:157:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__tuple:16:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:

----------------------------------------------------------------------------

/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:507:52: Template argument for template type parameter must be a type
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDNavigationElement.mm:31:9: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDNavigationElement.mm:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/nav_element.h:24:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/nav_element.h:24:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:12: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:17: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/detail/config_all.hpp:151:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/utility:21:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/utility:157:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/utility:157:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__tuple:16:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__tuple:16:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/type_traits:22:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/type_traits:500:17: Template parameter is declared here

----------------------------------------------------------------------------


/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/exception
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/exception:129:45: Field has incomplete type 'std::exception_ptr'
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDNavigationElement.mm:31:9: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDNavigationElement.mm:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/nav_element.h:24:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/nav_element.h:24:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/sstream:174:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ostream:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ostream:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:131:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/ostream:131:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ios:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ios:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:15:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__locale:15:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/string:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/string:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string:439:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/string:439:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/algorithm:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/algorithm:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:627:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:627:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/memory:15:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/memory:15:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:594:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:594:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/typeinfo:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/typeinfo:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/typeinfo:61:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/typeinfo:61:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/exception:25:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/exception:25:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/exception:124:24: Definition of 'std::exception_ptr' is not complete until the closing '}'
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/exception:129:55: Expected ';' at end of declaration list
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDNavigationElement.mm:31:9: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/RDServices/Main/RDNavigationElement.mm:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/nav_element.h:24:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/nav_element.h:24:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/epub3.h:25:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:10: In file included from /Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/Platform/Apple/include/ePub3/utilities/basic.h:31:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/sstream:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/sstream:174:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/sstream:174:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ostream:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ostream:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ostream:131:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/ostream:131:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ios:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/ios:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/ios:216:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/ios:216:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/__locale:15:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/__locale:15:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/string:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/string:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/string:439:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/string:439:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/algorithm:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/algorithm:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/algorithm:627:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/algorithm:627:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/memory:15:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/memory:15:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/memory:594:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/memory:594:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/typeinfo:18:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/typeinfo:18:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/typeinfo:61:10: In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/typeinfo:61:
/Users/jim/Documents/Xcode/shelfit-ios/readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/exception:25:20: In file included from readium-sdk/ePub3/ThirdParty/boost/include/boost/tr1/tr1/exception:25:


----------------------------------------------------------------------------

**Comment by AlanQuatermain** [comment]
It looks to me as though the Boost TR1 type_traits and the stdlib type_traits are getting confused. Where is the boost one coming from? That should only be used on Android IIRC, to fill the gaps the GNU folks left all over their own C++11 stdlib implementation.

**Comment by jimjeffers** [comment]
@AlanQuatermain - You're guess is as good as mine! It was not included in the actual XCode project so I'm not sure why or how it was getting pulled in. At first I simply tried deleting the block from _config.h:

#if EPUB_OS(ANDROID)
//# define UTF_USE_ICU 1
# define CXX11_STRING_UNAVAILABLE 1
# undef REGEX_INCLUDE
# define REGEX_INCLUDE <boost/regex.hpp>
# undef REGEX_NS
# define REGEX_NS boost
# undef LOCALE_INCLUDE
# define LOCALE_INCLUDE <boost/locale.hpp>
# undef LOCALE_NS
# define LOCALE_NS boost
# if EPUB_COMPILER(CLANG)
#  define nan(x) __builtin_nan(x)
# endif
#endif

But that made no difference so I just undid that change. So I went into the actual project and deleted the boost directory from disk. Now the project compiles just fine! Really confusing, I can't see any clear reason boost was getting included, let alone how...

Now that I'm able to compile I'll keep moving forward w/ integration. Do you guys think the boost issue might be worth looking into as an actual issue?

**Comment by AlanQuatermain** [comment]
It might be something very simple— if you dragged the source into a new Xcode project then likely it would by default include all files in the main target, thus adding the Boost headers to the default search paths. That would be my guess, anyway.

I had thought there was an OS X framework target in the ePub3 Xcode project already; if so, looking at that may indicate the differences. This was all prior to iOS 8 allowing actual frameworks in applications though, hence the iOS target building a static library.

FWIW, the static library approach may work out somewhat better, especially for heavily template-based code. When the system links the static library it can prune a *lot* of unused things, which it can’t do with a dynamic library since there’s no definitive ‘link time’ at which it can determine whether any of the code paths are unused. If you’re not actually using the framework as a dynamic piece of multiple items (i.e. you have an extension residing in your app bundle, where both the app & the extension need to use the same code) then I’d suggest trying the static library approach anyway, since it can potentially make quite a bit of difference.

Anecdote: I once had to include AES-GCM support in an ObjC framework, and the only (free) implementation I could find was in a 26MB library. Compiling it as a static lib though, the end result imported about 1MB from that library and discarded all the rest, IIRC.

**Comment by jimjeffers** [comment]
@AlanQuatermain You called it. The project had `$(PROJECT_DIR)/readium-sdk/ePub3` as a recursive search path. I removed it and the project compiles. Also, I had to add `$(PROJECT_DIR)/readium-sdk/Platform/Apple` to my header search paths to get RDServices. I notice the SDKLauncher just has `readium-sdk/ePub3` in the header search paths. Not sure what magic is going on here but it seems to compile and ignore boost just fine now.

**Comment by jimjeffers** [comment]
Hey guys,
Sorry for leaving this open. This discussion helped me resolve all of the compiler issues that were blocking me when integrating the SDK Launcher into my own app. Once this project is done I'll write up a retrospective with some basic docs to help anyone replicate the settings/gotchas to integrate readium into their own existing projects. Thanks for all of your help here!

**Comment by danielweck** [comment]
That would be great @jimjeffers thanks! :)
